### PR TITLE
Fix Lambda rate limit by backoff and retry when exceeding

### LINKED ISSUE #2261: [Lambda] Handle rate limit error crashing `sky launch`
User reported:
> For lambda, I'm getting api rate limit errors. 
> However, the problem is that I'm guessing there are some secondary api calls that causes sky launch to fail because of the rate limi, so the job never gets sent, but the node itself gets provisioned. 
> And because sky launch fails, it doesn't autodown, so it's just left running

Error:
```
...
File "/tmp/skypilot_ray_up_niigobvr.py", line 76, in <module>
sdk.create_or_update_cluster('/root/.sky/generated/training-228-53355c15.yml', **{'no_restart': True})
File "/usr/local/lib/python3.8/site-packages/ray/autoscaler/sdk/sdk.py", line 38, in create_or_update_cluster
return commands.create_or_update_cluster(
File "/usr/local/lib/python3.8/site-packages/ray/autoscaler/_private/commands.py", line 282, in create_or_update_cluster
get_or_create_head_node(
File "/usr/local/lib/python3.8/site-packages/ray/autoscaler/_private/commands.py", line 726, in get_or_create_head_node
provider.create_node(head_node_config, head_node_tags, 1)
File "/usr/local/lib/python3.8/site-packages/sky/skylet/providers/lambda_cloud/node_provider.py", line 287, in create_node
vms = self._list_instances_in_cluster()
File "/usr/local/lib/python3.8/site-packages/sky/skylet/providers/lambda_cloud/node_provider.py", line 127, in _list_instances_in_cluster
vms = self.lambda_client.list_instances()
File "/usr/local/lib/python3.8/site-packages/sky/skylet/providers/lambda_cloud/lambda_utils.py", line 160, in list_instances
raise_lambda_error(response)
File "/usr/local/lib/python3.8/site-packages/sky/skylet/providers/lambda_cloud/lambda_utils.py", line 78, in raise_lambda_error
raise LambdaCloudError('Your API requests are being rate limited.')
sky.skylet.providers.lambda_cloud.lambda_utils.LambdaCloudError: Your API requests are being rate limited.
```

### PR DESCRIPTION
**Pull request by cblmemo**
This PR fixes #2261 by:
* Add retry with backoff to all Lambda API requests when the rate limit exceeds;
* Retry launching when exceeding the rate limit.

TODO: debug and verify it works properly

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

**Comment by cblmemo** [comment]
Tried `for i in $(seq 5); do sky launch --cpus 1+ --cloud lambda --down -y -c lambda-$i &; done`, 4 of them launched successfully and 1 of them failed due to ssh connection timeout:

```bash
I 07-24 11:07:12 cloud_vm_ray_backend.py:911] ====== stdout ======
2023-07-24 11:06:34,497 INFO commands.py:276 -- Cluster: lambda-5
2023-07-24 11:06:34,497 VWARN commands.py:324 -- Loaded cached provider configuration from /tmp/ray-config-3959b5f8e1d478f99ddf4ae794347a98c33064b0
2023-07-24 11:06:34,497 WARN commands.py:332 -- If you experience issues with the cloud provider, try re-running the command with --no-config-cache.
E 07-24 11:07:11 subprocess_utils.py:73] ssh: connect to host 192.9.139.61 port 22: Connection timed out
E 07-24 11:07:11 subprocess_utils.py:73] 

I 07-24 11:07:12 cloud_vm_ray_backend.py:914] ====== stderr ======
Dropping the empty legacy field head_node. head_nodeis not supported for ray>=2.0.0. It is recommended to removehead_node from the cluster config.
Dropping the empty legacy field worker_nodes. worker_nodesis not supported for ray>=2.0.0. It is recommended to removeworker_nodes from the cluster config.
sky.exceptions.CommandError: Command ip -4 -br addr show | grep UP | grep -Eo "(10\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)|172\.(1[6-9]|2[0-9]|3[0-1]))\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)" failed with return code 255.
Failed get obtain private IP from node
```

Maybe I can fix this problem in another PR since it is not relative to the rate limit. I think this PR is ready for another look 🙂 @concretevitamin

File: sky/skylet/providers/lambda_cloud/lambda_utils.py [hunk 1]
```diff
@@ -2,10 +2,17 @@
...
 import requests
+import time
 from typing import Any, Dict, List, Optional, Tuple
 
+from sky.utils import common_utils
+
 CREDENTIALS_PATH = '~/.lambda_cloud/lambda_keys'
 API_ENDPOINT = 'https://cloud.lambdalabs.com/api/v1'
+# TODO(tian): Determine best backoff factors.
+INITIAL_BACKOFF = 3
```
Comment (concretevitamin) [inline sky/skylet/providers/lambda_cloud/lambda_utils.py:13]: I tried the repro and for the first ~3 minutes, no launch can go through:
```
I 07-24 20:59:38 cloud_vm_ray_backend.py:1732] Launching on Lambda us-west-1
⠧ LaunchingI 07-24 20:59:38 cloud_vm_ray_backend.py:1732] Launching on Lambda us-west-1
⠇ LaunchingI 07-24 20:59:38 cloud_vm_ray_backend.py:1732] Launching on Lambda us-west-1
⠋ LaunchingI 07-24 20:59:38 cloud_vm_ray_backend.py:1732] Launching on Lambda us-west-1
⠹ LaunchingI 07-24 20:59:38 cloud_vm_ray_backend.py:1732] Launching on Lambda us-west-1
⠹ LaunchingI 07-24 21:00:18 cloud_vm_ray_backend.py:1793] Retrying due to Lambda API rate limit exceeded.
I 07-24 21:00:18 cloud_vm_ray_backend.py:1826] Retrying launching in 6.8 seconds.
⠦ LaunchingI 07-24 21:00:32 cloud_vm_ray_backend.py:1793] Retrying due to Lambda API rate limit exceeded.
I 07-24 21:00:32 cloud_vm_ray_backend.py:1826] Retrying launching in 5.1 seconds.
⠦ LaunchingI 07-24 21:00:39 cloud_vm_ray_backend.py:1793] Retrying due to Lambda API rate limit exceeded.
I 07-24 21:00:39 cloud_vm_ray_backend.py:1826] Retrying launching in 6.8 seconds.
⠋ LaunchingI 07-24 21:01:14 cloud_vm_ray_backend.py:1793] Retrying due to Lambda API rate limit exceeded.
I 07-24 21:01:14 cloud_vm_ray_backend.py:1826] Retrying launching in 4.0 seconds.
⠇ LaunchingI 07-24 21:01:24 cloud_vm_ray_backend.py:1793] Retrying due to Lambda API rate limit exceeded.
I 07-24 21:01:24 cloud_vm_ray_backend.py:1826] Retrying launching in 7.1 seconds.
I 07-24 21:01:24 cloud_vm_ray_backend.py:1793] Retrying due to Lambda API rate limit exceeded.
I 07-24 21:01:24 cloud_vm_ray_backend.py:1826] Retrying launching in 14.1 seconds.
⠧ LaunchingI 07-24 21:01:45 cloud_vm_ray_backend.py:1793] Retrying due to Lambda API rate limit exceeded.
I 07-24 21:01:45 cloud_vm_ray_backend.py:1826] Retrying launching in 7.8 seconds.
⠧ LaunchingI 07-24 21:02:00 cloud_vm_ray_backend.py:1793] Retrying due to Lambda API rate limit exceeded.
I 07-24 21:02:00 cloud_vm_ray_backend.py:1826] Retrying launching in 17.2 seconds.
⠹ Launching
```

Should we try increasing initial backoff to say like 10 seconds to lessen the contention?

**Review by concretevitamin — COMMENTED** [review]
Thanks @cblmemo! A question about backoff settings.

File: sky/skylet/providers/lambda_cloud/lambda_utils.py (hunk 1 above)
Comment (cblmemo) [reply sky/skylet/providers/lambda_cloud/lambda_utils.py]: Tested initial = 10s and it performs very well (not a single retry happened when launching 5 clusters simultaneously). I think 10s is a good choice.

File: sky/skylet/providers/lambda_cloud/lambda_utils.py [hunk 2]
```diff
@@ -92,25 +91,14 @@ def raise_lambda_error(response: requests.Response) -> None:
     raise LambdaCloudError(f'{code}: {message}')
 
 
-def try_post_with_backoff(url: str, data: str, headers: Dict[str, str]):
-    backoff = common_utils.Backoff(initial_backoff=INITIAL_BACKOFF,
+def _try_request_with_backoff(method,
```
Comment (concretevitamin) [inline sky/skylet/providers/lambda_cloud/lambda_utils.py:94]: : Callable[...]? Or we can pass a str.

**Review by concretevitamin — APPROVED** [review]
Nice! Thanks for fixing this @cblmemo.

File: sky/skylet/providers/lambda_cloud/lambda_utils.py (hunk 2 above)
Comment (cblmemo) [reply sky/skylet/providers/lambda_cloud/lambda_utils.py]: str sounds good 👌🏻

**Pull request merged**